File: code/go/ch18/src/luago/state/api_misc.go
package state

import "luago/number"

// [-0, +1, e]
// http://www.lua.org/manual/5.3/manual.html#lua_len
func (self *luaState) Len(idx int) {
	val := self.stack.get(idx)

	if s, ok := val.(string); ok {
		self.stack.push(int64(len(s)))
	} else if result, ok := callMetamethod(val, val, "__len", self); ok {
		self.stack.push(result)
	} else if t, ok := val.(*luaTable); ok {
		self.stack.push(int64(t.len()))
	} else {
		panic("length error!")
	}
}

// [-n, +1, e]
// http://www.lua.org/manual/5.3/manual.html#lua_concat
func (self *luaState) Concat(n int) {
	if n == 0 {
		self.stack.push("")
	} else if n >= 2 {
		for i := 1; i < n; i++ {
			if self.IsString(-1) && self.IsString(-2) {
				s2 := self.ToString(-1)
				s1 := self.ToString(-2)
				self.stack.pop()
				self.stack.pop()
				self.stack.push(s1 + s2)
				continue
			}

			b := self.stack.pop()
			a := self.stack.pop()
			if result, ok := callMetamethod(a, b, "__concat", self); ok {
				self.stack.push(result)
				continue
			}

			panic("concatenation error!")
		}
	}
	// n == 1, do nothing
}

// [-1, +(2|0), e]
// http://www.lua.org/manual/5.3/manual.html#lua_next
func (self *luaState) Next(idx int) bool {
	val := self.stack.get(idx)
	if t, ok := val.(*luaTable); ok {
		key := self.stack.pop()
		if nextKey := t.nextKey(key); nextKey != nil {
			self.stack.push(nextKey)
			self.stack.push(t.get(nextKey))
			return true
		}
		return false
	}
	panic("table expected!")
}

// [-1, +0, v]
// http://www.lua.org/manual/5.3/manual.html#lua_error
func (self *luaState) Error() int {
	err := self.stack.pop()
	panic(err)
}

// [-0, +1, –]
// http://www.lua.org/manual/5.3/manual.html#lua_stringtonumber
func (self *luaState) StringToNumber(s string) bool {
	if n, ok := number.ParseInteger(s); ok {
		self.PushInteger(n)
		return true
	}
	if n, ok := number.ParseFloat(s); ok {
		self.PushNumber(n)
		return true
	}
	return false
}

File: code/csharp/ch13/luavm/state/ApiMisc.cs
using System;

namespace luavm.state
{
    public partial class LuaState
    {
        public void Len(int idx)
        {
            var val = _stack.Get(idx);
            if (val is string str)
            {
                PushInteger(str.Length);
                return;
            }

            var (result, ok) = CallMetamethod(val, val, "__len", this);
            if (ok)
            {
                _stack.Push(result);
                return;
            }

            if (!(val is LuaTable table)) throw new Exception("length error!");
            PushInteger(table.Len());
            return;
        }

        public void Concat(int n)
        {
            if (n == 0)
            {
                _stack.Push("");
            }
            else if (n >= 2)
            {
                for (var i = 1; i < n; i++)
                {
                    if (IsString(-1) && IsString(-2))
                    {
                        var s2 = ToString(-1);
                        var s1 = ToString(-2);
                        _stack.Pop();
                        _stack.Pop();
                        _stack.Push(string.Concat(s1, s2));
                        continue;
                    }

                    var b = _stack.Pop();
                    var a = _stack.Pop();
                    var (result, ok) = CallMetamethod(a, b, "__concat", this);
                    if (ok)
                    {
                        _stack.Push(result);
                        continue;
                    }


                    throw new Exception("concatenation error!");
                }

                // n==1, do nothing
            }
        }

        public uint RawLen(int idx)
        {
            var val = _stack.Get(idx);
            switch (val)
            {
                case string valStr:
                    return (uint) valStr.Length;
                case LuaTable luaTable:
                    return (uint) (luaTable).Len();
                default:
                    return 0;
            }
        }

        public bool Next(int idx)
        {
            var val = _stack.Get(idx);
            if (!(val is LuaTable t)) throw new Exception("table expected!");
            var key = _stack.Pop();
            var nextKey = t.NextKey(key);
            if (nextKey == null) return false;
            _stack.Push(nextKey);
            _stack.Push(t.Get(nextKey));
            return true;
        }

        public int Error()
        {
            var err = _stack.Pop();
            throw new Exception(Convert.ToString(err));
        }
    }
}